kubernetes/apps_kubeclient.go

```diff
@@ -770,63 +867,433 @@ func (kc *kubeClient) getPods(namespace string, uid types.UID) ([]v1.Pod, error)
-			appPods = append(appPods, pod)
+			appPods = append(appPods, &pod)
 		}
 	}
 
 	return appPods, nil
 }
 
-func (kc *kubeClient) getPodStatus(pods []v1.Pod) (*app.PodStats, error) {
-	var starting, running, stopping int
+// Pod status constants
+const (
+	podRunning     = "Running"
+	podNotReady    = "Not Ready"
+	podWarning     = "Warning"
+	podError       = "Error"
+	podPulling     = "Pulling"
+	podPending     = "Pending"
+	podSucceeded   = "Succeeded"
+	podTerminating = "Terminating"
+	podUnknown     = "Unknown"
+)
+
+func (kc *kubeClient) getPodStatus(pods []*v1.Pod) ([][]string, int) {
 	/*
-	 * TODO Logic for pod phases in web console is calculated in the UI:
+	 * Use the same categorization used by the web console. See:
 	 * https://github.com/openshift/origin-web-console/blob/v3.7.0/app/scripts/directives/podDonut.js
 	 * https://github.com/openshift/origin-web-console/blob/v3.7.0/app/scripts/filters/resources.js
-	 * Should we duplicate the logic here in Go, opt for simpler phases (perhaps just PodPhase), or send Pod as JSON to fabric8-ui
-	 * to reuse JS components
-	 * const phases = []string{"Running", "Not Ready", "Warning", "Error", "Pulling", "Pending", "Succeeded", "Terminating", "Unknown"}
 	 */
+	podStatus := make(map[string]int)
+	podTotal := 0
 	for _, pod := range pods {
-		// Terminating pods have a deletionTimeStamp set
-		if pod.ObjectMeta.DeletionTimestamp != nil {
-			stopping++
-		} else if pod.Status.Phase == v1.PodPending {
-			// TODO Is this a good approximation of "Starting"?
-			starting++
-		} else if pod.Status.Phase == v1.PodRunning {
-			running++
+		statusKey := podUnknown
+		if pod.Status.Phase == v1.PodFailed {
+			// Failed pods are not included, see web console:
+			// https://github.com/openshift/origin-web-console/blob/v3.7.0/app/scripts/directives/podDonut.js#L32
+			continue
+		} else if pod.DeletionTimestamp != nil {
+			// Terminating pods have a deletionTimeStamp set
+			statusKey = podTerminating
+		} else if warn, severe := isPodWarning(pod); warn {
+			// Check for warnings/errors
+			if severe {
+				statusKey = podError
+			} else {
+				statusKey = podWarning
+			}
+		} else if isPullingImage(pod) {
+			// One or more containers is waiting on its image to be pulled
+			statusKey = podPulling
+		} else if pod.Status.Phase == v1.PodRunning && !isPodReady(pod) {
+			// Pod is running, but one or more containers is not yet ready
+			statusKey = podNotReady
 		} else {
-			// TODO Handle other phases
+			// Use Kubernetes pod phase
+			statusKey = string(pod.Status.Phase)
+		}
+		podStatus[statusKey]++
+		podTotal++
+	}
+
+	// if there were actually no pods, create a dummy entry
+	if podTotal == 0 {
+		podStatus[podRunning] = 0
+	}
+
+	var result [][]string
+	for status, count := range podStatus {
+		statusEntry := []string{status, strconv.Itoa(count)}
+		result = append(result, statusEntry)
+	}
+
+	return result, podTotal
+}
+
+func isPodWarning(pod *v1.Pod) (warning, severe bool) {
```

Comment:
is this method covered by 1 or more tests ?

Reply:
A whitebox test has not yet been written.